src/array.rs

```diff
@@ -251,28 +252,30 @@ impl<T, D> PyArray<T, D> {
 }
 
 impl<T: Element, D: Dimension> PyArray<T, D> {
-    fn extract<'py, E>(ob: &'py PyAny) -> Result<&'py Self, E>
+    fn extract<'a, 'py, E>(ob: &'a Bound<'py, PyAny>) -> Result<&'a Bound<'py, Self>, E>
     where
-        E: From<PyDowncastError<'py>> + From<DimensionalityError> + From<TypeError<'py>>,
+        E: From<DowncastError<'a, 'py>> + From<DimensionalityError> + From<TypeError<'a>>,
     {
         // Check if the object is an array.
         let array = unsafe {
             if npyffi::PyArray_Check(ob.py(), ob.as_ptr()) == 0 {
-                return Err(PyDowncastError::new(ob, Self::NAME).into());
+                return Err(DowncastError::new(ob, <Self as PyTypeInfo>::NAME).into());
             }
-            &*(ob as *const PyAny as *const Self)
+            ob.downcast_unchecked()
         };
 
+        let arr_gil_ref: &PyArray<T, D> = array.as_gil_ref();
```

Comment:
Why do we need a GIL ref in the implementation here?

Reply:
The methods are not yet converted here, so we have nothing available on `Bound<'_, PyArray<T, D>>`. This will go away once we have those in place.